packages/ai-jsx/src/batteries/constrained-output.tsx

```diff
@@ -5,169 +5,255 @@
-      {children}
-    </ObjectFormatChatCompletion>
+export async function* YamlChatCompletion(
+  props: Omit<TypedObjectCompletionWithRetry, 'typeName' | 'parser'>,
+  { render }: AI.ComponentContext
+) {
+  return yield* render(
+    <ObjectCOmpletionWithRetry {...props} typeName="YAML" parser={yaml.load as (str: string) => object} />
   );
 }
 
 /**
  * A {@link ChatCompletion} component that constrains the output to be a valid object format (e.g. JSON/YAML).
  *
- * @param retries The maximum number of times to retry the completion if the output is invalid.
- * @param validator A function that returns a {@link ValidationResult} for a given string.
- * @param typeName The name of the type, to be used in the prompt.
- * @param children The children to render.
- * @returns A string that validates as the given type or throws an error after `retries` attempts
+ * Though not required, you can provide a Zod schema to validate the output against. This is useful for
+ * ensuring that the output is of the expected type.
+ *
+ * @returns A string that validates as the given type or throws an error.
+ *    Intermediate results that are valid, are also yielded.
  */
-async function ObjectFormatChatCompletion(
-  {
-    retries = 3,
-    validator,
-    typeName,
-    children,
-    ...props
-  }: {
-    validator: (str: string) => ValidationResult;
-    /** function that succeeds if output string is of the expected format */
-    typeName: string;
-    /** name of the type, to be used in the prompt */
-    retries?: number;
-    /** retries if output fails validation, how many times should we retry.
-     * Note: you can have `1 + retries` LLM calls in total. */
-    children: AI.Node;
-  },
+async function* OneShotObjectCompletion(
+  { children, typeName, validators, example, schema, parser, partialResultCleaner, ...props }: TypedObjectCompletion,
   { render, logger }: AI.ComponentContext
 ) {
+  // If a schema is provided, it is added to the list of validators as well as the prompt.
+  const validatorsAndSchema = schema ? [schema.parse, ...(validators ?? [])] : validators ?? [];
+
   const childrenWithCompletion = (
     <ChatCompletion {...props}>
       {children}
       <SystemMessage>
-        Your response must be a valid {typeName}. Do not include ```{typeName.toLowerCase()} ``` code blocks.
+        Respond with a {typeName} object that encodes your response.
+        {schema
+          ? `The ${typeName} object should match this JSON Schema: ${JSON.stringify(zodToJsonSchema(schema))}`
+          : ''}
+        {example ? `For example: \n${example}` : ''}
+        Respond with only the {typeName} object. Do not include any explanatory prose. Do not include ```
+        {typeName.toLowerCase()} ``` code blocks.
       </SystemMessage>
     </ChatCompletion>
   );
+  const renderGenerator = render(childrenWithCompletion)[Symbol.asyncIterator]();
+
+  // TODO: we can possibly add a timelimit here so we don't emit too many times.
```

Comment:
I'm worried that calling parser, schema, and all the validators for each returned result can be a bit too much. We can potentially add a time limit for it too.

Reply:
Unless I'm misunderstanding what those functions do, I strongly suspect their overhead will be trivial relative to the model latency.

Reply:
Yeah I think so too.